### ISSUE TITLE
Fetch Current Schema and Table for Granting Permissions in Post-Hooks

### ISSUE THREAD

**Issue by jrhorne** [TobikoData/sqlmesh#2974]
As a data transformer, I would like to be able to grant permissions when new schemas and tables are created. For example, I have a role in Postgres called data_consumer_admin.

Regardless of the plan being executed, I would like the data_consumer_admin to be able to access the newly created schemas and tables. For example, to review MRs/PRs, I share an example of the data processed. Every data_consumer_admin should be able to query from this SQLMesh environment.

This runs, but marts is not replaced with marts__dev or whatever the current plan's environment is.

```
@IF(
  @runtime_stage = 'creating',
  GRANT USAGE ON SCHEMA marts TO group_data_consumer_admin
);
```

From Slack discussion here: [https://tobiko-data.slack.com/archives/C044BRE5W4S/p1722431705430829](https://tobiko-data.slack.com/archives/C044BRE5W4S/p1722431705430829)

**Comment by georgesittas** [comment]
Hey @jrhorne, after discussing about this internally we concluded that exposing the current plan's environment to the macro evaluator is not the right way to solve this problem. You can instead leverage the [Python API](https://sqlmesh.readthedocs.io/en/stable/reference/python/) to directly `execute` statements that grant permissions after you run the plan.

For example, after creating a new "dev" environment, you could do the following for postgres, assuming you have >1 schemas and want to automate the permission granting (otherwise you could simply hardcode the schema name):

```python
from sqlmesh.core.context import Context

ctx = Context()
engine_adapter = ctx.engine_adapter
dev_schemas = engine_adapter.fetchdf(
    "select schema_name from information_schema.schemata where schema_name like '%__dev'"
)
for dev_schema in dev_schemas:
    engine_adapter.execute(f"GRANT USAGE ON SCHEMA {dev_schema} TO group_data_consumer_admin")
```

I'm closing this, but happy to continue discussing.

CC: @izeigerman

**Comment by jrhorne** [comment]
Makes sense. Here's my script assign_dev_permissions.py if anyone stumbles upon this later:

`"""
README:

Purpose:
--------
tl;dr Run this script to give other DS members access to testing schemas for reviewing data changes

This script automates the assignment of specific permissions to designated roles in development schemas.
The script targets schemas in a database that contain the pattern '__dev' in their names, commonly used for development and testing environments.

The script iterates over each of these schemas and assigns the following permissions to each role listed in `roles_to_assign`:

Roles and Permissions:
----------------------
1. `group_data_consumer_admin`:
   - Granted USAGE on the schema, allowing access to the schema without direct access to its contents.
   - Granted SELECT on all tables in the schema, enabling read-only access to data.
   - Granted EXECUTE on all functions in the schema, allowing the execution of stored procedures and functions.
   - Granted USAGE on all sequences in the schema, allowing access to the sequence objects, typically for reading their current value.

Prerequisites:
--------------
- Ensure that the database connection and context are correctly set up using `sqlmesh`.
- Verify that the user running this script has the necessary privileges to grant the described permissions in the database.

Instructions:
-------------
1. Install the required package (`sqlmesh`).
2. Run this script in an environment with the necessary privileges.

"""

from sqlmesh.core.context import Context

# Initialize the context and retrieve the engine adapter
ctx = Context()
engine_adapter = ctx.engine_adapter

# Fetch all development schemas that match the naming pattern
dev_schemas = engine_adapter.fetchdf(
    "SELECT schema_name FROM information_schema.schemata WHERE schema_name LIKE '%__%dev%'"
)

# List of roles to assign permissions to
roles_to_assign = ["group_data_consumer_admin"]

# Iterate over each role and assign the necessary permissions for each development schema
for role_to_assign in roles_to_assign:
    for dev_schema in dev_schemas.schema_name:
        engine_adapter.execute(
            f"GRANT USAGE ON SCHEMA {dev_schema} TO {role_to_assign}"
        )
        engine_adapter.execute(
            f"GRANT SELECT ON ALL TABLES IN SCHEMA {dev_schema} TO {role_to_assign}"
        )
        engine_adapter.execute(
            f"GRANT EXECUTE ON ALL FUNCTIONS IN SCHEMA {dev_schema} TO {role_to_assign}"
        )
        engine_adapter.execute(
            f"GRANT USAGE ON ALL SEQUENCES IN SCHEMA {dev_schema} TO {role_to_assign}"
        )

        print(f"Assigned permissions to {role_to_assign} on {dev_schema}")
`
